Add a corpus, fix a bunch of issues

savetheclocktower opened the pull request:
This is a big blob of changes and I'd gladly break them up into smaller parts if requested. I started fixing them all at once and didn't think to untangle it all until later.

Here's what's new:

* Added the corpus from [my fork of the other `tree-sitter-svelte`](https://github.com/savetheclocktower/tree-sitter-svelte-himujjal) and adjusted it as necessary.
* I was driven nuts by how repetitive it was to write queries to highlight the `@html` and `#each` and `/if` inside of various Svelte template constructs, so I gave them all node types of `block_start_tag`, `block_tag`, and `block_end_tag`. Naming is negotiable.
* Svelte recognizes `{expressions}` even when they're inside of quoted attribute values, so I had to supersede the HTML grammar's behavior to allow that.
* Svelte allows `<Foo.Bar />` tag names just like JSX, so dots are permitted when scanning for tag names.
* Snippet names used to be lumped in with `svelte_raw_text`, but the syntax for defining a snippet is easy enough to recognize, so it should have its own node type. (This makes `tags.scm` queries easier.)
* A big one, and one that I'd want a proper C programmer to look at: after some discussion in Discord, I tried to make the parser more tolerant of constructs like `{"}"}` and `{2 + /* foo } */ + 2}`. So the scanner detects more modes in which `}` would be a false positive — quoted strings, template strings, block comments — and skips past them.

These changes represent all the work I had to do to get this parser at feature parity with my previous fork. There are a few more fixes worth making, but this is a good start.

savetheclocktower committed:
Parse snippet names instead of consuming them as `svelte_raw_text`

savetheclocktower committed:
Do more work to detect false positive `}`s when scanning raw text.

savetheclocktower committed:
Restore `tag` macro

savetheclocktower committed:
Detect Svelte expressions in quoted attribute values

savetheclocktower committed:
Ignore `.DS_Store` files

savetheclocktower committed:
Add a corpus for Svelte constructs

savetheclocktower committed:
Rename `block_(open|close)_tag` to `block_(start|end)_tag` for consistency

savetheclocktower committed:
Move some stuff around in the corpus

savetheclocktower committed:
Tolerate complex content in `#snippet` params and `@render` arguments

savetheclocktower committed:
Allow an optional chain operator in `@render`.

savetheclocktower committed:
Enforce separation of `@tag` from value

savetheclocktower committed:
Don't try to enforce HTML rules about containment or element content…

…unless the usage is all lowercase.

Browsers treat HTML case-insensitively, but the Svelte compiler doesn't. `Input` and `input` are two different things, and the former does not have anything to do with the latter.

savetheclocktower committed:
Handle JavaScript line comments in `svelte_raw_text` blocks

savetheclocktower committed:
Don't try to generate a `snippet_name` node for `@render` tags…

…since their argument can take arbitrary JavaScript.

savetheclocktower committed:
Permit Svelte expression tags in quoted attribute values

savetheclocktower committed:
Permit Svelte statement blocks in quoted attribute values

savetheclocktower committed:
Add corpus tests for statements and expressions in attribute values

savetheclocktower commented:
I hadn't yet noticed the CI workflow and all the examples it ran against, so I took a pass at fixing the ones that don't pass. I think many of them weren't passing in the first place, but a few were easily fixable:

* Svelte enforces that HTML tags be lowercase even when browsers don't, since that's how it can distinguish between HTML and Svelte components in markup. `tree-sitter-html`'s validation logic (enforcing which tags can have children and which ones can go inside which others) needed to be changed to operate only on lowercase tags.
* Line comments needed to be handled — I'd thought they were flat-out invalid in plain expressions, but the Svelte playground tolerates a line comment as long as you've got a line break at the end to delimit it before the closing brace.

Meanwhile, a few are new and need to be added to `invalid-files`, and a couple involve errors generated by `tree-sitter-html` — hence the fix would need to be in that parser, or else those examples should also be moved to `invalid-files`.

This won't get CI green, but should get it closer, at least.

matubu commented on grammar.js:
```diff
@@ -49,19 +51,66 @@ module.exports = grammar(HTML, {
+        /[^\\{']+/,
+        // …or an expression.
+        $.expression,
+        $._svelte_expression,
+        $._svelte_statement,
```
I don't think `_svelte_expression` of `_svelte_statement` can be in attribute value

savetheclocktower replied:
Ha, you're right. I was trying to make CI happy and I didn't check [the context on this test case](https://github.com/sveltejs/svelte/blob/30b143cef0d5ea17a6edec33496681882eadce34/packages/svelte/tests/validator/samples/logic-block-in-attribute/errors.json). I'll roll this back. Thanks!

savetheclocktower committed:
Outlaw Svelte expressions/blocks in attribute values…

…because they're not allowed and I can't read.

savetheclocktower committed:
Revert backslash escape logic in quoted attributes…

…since they don't exist.

savetheclocktower committed:
Add another two examples to `invalid-files`

savetheclocktower committed:
Add three more examples to `invalid-files`…

…for reasons I'll explain in the PR, though it would be nice to be able to annotate an entry in the YAML file with the reason why it ought to fail.

savetheclocktower commented:
So I think the remainder of the CI failures can be explained by new examples in the `svelte` repo. I went through them, figured out whether they should pass or fail (after a few false starts), then either fixed the issues or added them to `invalid-files` as appropriate.

Of the new entries to `invalid-files`, the only controversial ones might be two examples (`each-updates-6` and `each-updates-7`) that contain an `ERROR` node because the examples use an unencoded ampersand. Since `tree-sitter-html` behaves the same way, I decided those failures were intentional. (It might not be obvious why they're supposed to fail to the next person who contributes. I didn't see any way of leaving a note next to the entry in the workflow file, but that might be a nice future enhancement.)

Rebase incoming.

savetheclocktower committed:
Regenerate after rebase

savetheclocktower committed:
Make `then_block` and `catch_block` equivalent to `else_if_block`…

…in their structure. Introduce `then_start` and `catch_start` to fill the role that `then_block` and `catch_block` were previously filling.

savetheclocktower committed:
Regenerate parser with the bundled version of `tree-sitter-cli`

savetheclocktower committed:
Rephrase scanner comment

savetheclocktower committed:
Add scanner coverage

savetheclocktower committed:
Remove unnecessary `$` parameter from `tag`

savetheclocktower committed:
Regenerate with 0.23.0

AlbertMarashi commented:
Great job. This looks better
